include/oneapi/dpl/pstl/utils.h

```diff
@@ -818,30 +822,57 @@ __shars_upper_bound(_Acc __acc, _Size __first, _Size __last, const _Value& __val
                                    oneapi::dpl::__internal::__reorder_pred<_Compare>{__comp}});
 }
 
-// TODO In C++20 we may try to use std::equality_comparable
+#if _ONEDPL_CPP20_CONCEPTS_PRESENT
+
+template <typename _Iterator1, typename _Iterator2>
+inline constexpr bool __is_equality_comparable_with_v = std::equality_comparable_with<_Iterator1, _Iterator2>;
+
+#else
+
 template <typename _Iterator1, typename _Iterator2, typename = void>
-struct __is_equality_comparable : std::false_type
+struct __has_equality_op : std::false_type
 {
 };
 
-// All with implemented operator ==
 template <typename _Iterator1, typename _Iterator2>
-struct __is_equality_comparable<
-    _Iterator1, _Iterator2,
-    std::void_t<decltype(::std::declval<::std::decay_t<_Iterator1>>() == ::std::declval<::std::decay_t<_Iterator2>>())>>
+struct __has_equality_op<_Iterator1, _Iterator2,
+                         std::void_t<decltype(std::declval<_Iterator1>() == std::declval<_Iterator2>())>>
     : std::true_type
 {
 };
 
+template <typename _Iterator1, typename _Iterator2>
+struct __is_equality_comparable_with : __has_equality_op<_Iterator1, _Iterator2>
+{
+};
+
+template <typename _Iterator1, typename _Iterator2>
+struct __is_equality_comparable_with<std::reverse_iterator<_Iterator1>, std::reverse_iterator<_Iterator2>>
+    : __is_equality_comparable_with<_Iterator1, _Iterator2>
+{
+};
+
+template <typename _Iterator1, typename _Iterator2>
+struct __is_equality_comparable_with<std::move_iterator<_Iterator1>, std::move_iterator<_Iterator2>>
```

Comment:
The question is the same as in our private conversation. Do we want to add extra layer with `decay_t`? our previous logic did that for iterators. On the idea level it should be like

```cpp
template <typename _Iterator1, typename _Iterator2>
struct __is_equality_comparable_with : __is_equality_comparable_with_impl<std::decay_t<_Iterator1>, std::decay_t<_Iterator2>>
{
};
```

and then we have partial specializations for reverse and move iterators for `_impl`

Reply:
Probably for the comparison with
```C++
template <typename _T>
struct __is_type_with_iterator_traits<
    _T, std::void_t<typename std::iterator_traits<std::remove_reference_t<_T>>::difference_type>> : std::true_type
{
};
```
and taking into account the discussion in the PR https://github.com/uxlfoundation/oneDPL/pull/2376 (https://github.com/uxlfoundation/oneDPL/pull/2376/files/c2ad9512c576d965561d12de72ba5d5a47b43b6c#r2267265672) I think we may have some common `iterator type` cleaner for both these usages like:
```C++
template <typename _Iterator>
using __clean_iterator_t = std::remove_reference_t<_Iterator>;
```
and use it.

Reply:
The trait you are talking about it different. Yes, there should not be a difference between reading (some) iterator traits from iterator and `const` iterator (of the same type). However, here you have a partial template specialization. If some how (during our manipulation with the type you end up with `const` iterator the partial specialization would not match and the code will go to a primary template. Thus just having `remove_reference` is not sufficient.

Reply:
Fixed

Reply:
I already approved, but I realized it would be nice to add a test or two checking that iterator types with `const` and/or different value categories work.  I don't believe that is covered in the tests.

Reply:
@danhoeflinger, done.